sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+
+```javascript
+function stubRecorder(pathToJson) {
+  const xhrData = []; // an empty array to hold the data
+  cy.server({ // if recording, save the response data in the array
+    onResponse: (response) => {
+      const url = response.url;
+      const method = response.method;
+      const data = response.response.body;
+      // We push a new entry into the xhrData array
+      xhrData.push({ url, method, data });
+    }
+  });
+  // this is used as a ‘filter’ for the data you want to record.
+  // You can specify the methods and routes that are recorded
+  cy.log('recording!');
+  cy.route({
+    method: 'GET',
+    url: '*',
+  });
+  // if recording, after the test runs, create a fixture file with the recorded data
+  after(function () {
+    cy.writeFile(`./cypress/fixtures/${pathToJson}.json`, xhrData);
+    cy.log(`Wrote ${xhrData.length} XHR responses to local file ${pathToJson}.json`);
+  });
+}
+
+/** Plays recorded fixture with all required network data as json*/
+function playStubbedFixture(stateFixture) {
+  cy.log(`playing fixture from ${stateFixture}`);
+  cy.fixture(stateFixture, { timeout: 15000 }) // the fixture file may be large and take time in CI
+    .then(data => {
+      for (let i = 0, length = data.length; i < length; i++) {
+        cy.route(data[i].method, data[i].url, data[i].data);
+      }
+    }).as(`stateFixture_stub`);
+}
+
+/** Visits the stubbed state */
+function visitStubbedState(stubFile, url, wait: boolean = true) {
+  playStubbedFixture(stubFile);
+  cy.visit(url);
+  if (wait) { // sometimes you do not want to wait for network, this gives you the option
+    cy.wait('@stateFixture_stub', { timeout: 15000 });
+  }
+}
+
+//////////
+// usage
+
+// recording network
+it('should run your test', function () {
+  stubrecorder('jsonfileNameForNetworkData');
+
+  // your original test
+
+  cy.wait(5000); // one-time wait so that the after() step records all the network without missing anything
+
+  // the rest of your original test
+});
+
+// playing the stubbed network
+it('should run your test', function () {
+  // every time we visit this endpoint, all network will be stubbed
+  // double check this by observing (XHR stubbed) network responses in the test runner
+  visitStubbedState('jsonfileNameForNetworkData', '/endpoint');
+
+  // the rest of your original test
+});
+```
+
+
+<br/><br/>
+
+### Application actions
+
+Cypress gives complete control of your application. You can bypass the page object abstraction layer (which is detached from your application) access the UI directly via `cy.get()` , have access to the API and database, and even access the source code.
+
+Component testing is an experimental feature in Cypress as of version 4.7.0, albeit some of the functionality is available. Refer to component testing articles under tools section for experiments with React and Storybook on this topic.
```

Comment:
It's not clear to me how it's correlated to the topic of the chapter 🤔

Reply:
We are talking about components, not necessarily component testing. Having control over them can allow to set the state of the application. 

Are you referring to the mention of "experimental feature"? I think we should let the reader know that this is not the official way, yet. We do not know what the official component testing setup is, but when we do, this information may become outdated. 

If it's super confusing, you can remove the excess parts.

Reply:
I don't get why component testing is cited here, what I mean:

- app actions allow you to short-circuit the UI, giving Cypress access to functions that skip the UI interactions and cuts the test duration
- app actions are often exposed by components

but
- component testing does not need app actions usually, or not so much, because the interactions made in a single component are limited. App actions are useful for E2E/Integration tests where the states of the app are a lot
- E2E testing and component testing cannot be mixed (in the same test I mean)

so, again, I don't get why component testing is cited here 🤔 (don't get me wrong, I can't understand if it's good or bad because I don't get the point at all at the moment 🙏)

Reply:
removed the reference to component testing